server/container/util/src/main/java/org/apache/james/util/AuditTrail.java

```diff
@@ -0,0 +1,91 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.util;
+
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableMap;
+
+public class AuditTrail {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AuditTrail.class);
+
+    public static Entry entry() {
+        return new Entry();
+    }
+
+    public static class Entry {
+        Optional<String> username = Optional.empty();
+        Optional<String> remoteIP = Optional.empty();
+        Optional<String> userAgent = Optional.empty();
+        Optional<String> protocol = Optional.empty();
+        Optional<String> action = Optional.empty();
+        Map<String, String> parameters = ImmutableMap.of();
+
+        private Entry() {
+
+        }
+
+        public Entry username(String username) {
```

Comment:
one more method will be less boilerplate?
`public Entry username(Username username)` 
`public Entry remoteIP(InetAddress remoteIP)`
...

Reply:
> public Entry username(Username username)

I am not sure if I should add `james-core` module into `james-server-util` module. But I will try.

Reply:
> I am not sure if I should add james-core module into james-server-util module.

Circuit dependency. Sorry on this...

Reply:
yes, ignored it